packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js

```diff
@@ -0,0 +1,154 @@
+const Web3 = require("web3");
+const { create, all } = require("mathjs");
+
+// Customize math (will be exported for other modules to use).
+const math = create(all, { number: "BigNumber", precision: 100 });
+const nativeIsAlpha = math.parse.isAlpha;
+const allowedSpecialCharacters = Array.from("[]/ -");
+
+// Modifies math.js's function to determine whether a character is allowed in a symbol name.
+math.parse.isAlpha = function(c, cPrev, cNext) {
+  // This character is the escape and the next is the special character.
+  const isValidEscapeChar = c === "\\" && allowedSpecialCharacters.includes(cNext);
+
+  // This character is the special character and the previous is the escape.
+  const isSpecialChar = cPrev === "\\" && allowedSpecialCharacters.includes(c);
+
+  return nativeIsAlpha(c, cPrev, cNext) || isValidEscapeChar || isSpecialChar;
+};
+
+// This escapes all the special characters (defined in the allowedSpecialCharacters array) in a string.
+// For example (note that "\\" is the js representation for a single literal backslash)
+// "ab/c d [e]" -> "ab\\/c\\ d\\[e\\]"
+function escapeSpecialCharacters(input) {
+  return Array.from(input)
+    .map((char, index, array) => {
+      if (allowedSpecialCharacters.includes(char) && array[index - 1] !== "\\") {
+        return `\\${char}`;
+      } else {
+        return char;
+      }
+    })
+    .join("");
+}
+
+// Allows users to combine other price feeds using "expressions" with the price feed identifiers being the symbols
+// in these expressions. Ex: "USDETH * COMPUSD". Users can also comfigure custom price feeds in their configuration
+// with custom symbols. Ex: "USDETH * COMPUSD * MY_CUSTOM_FEED".
+class ExpressionPriceFeed extends PriceFeedInterface {
+  /**
+   * @notice Constructs the DominationFinancePriceFeed.
+   * @param {Object} priceFeedMap an object mapping from price feed names to the price feed objects.
+   *                 Ex:
+   *                 {
+   *                   "USDETH": ETHBTC_PRICE_FEED_INSTANCE,
+   *                   "COMPUSD": COMPUSD_PRICE_FEED_INSTANCE
+   *                 }
+   * @param {string} expression a string expression that uses price feeds in the priceFeedMap to compute a price.
+   *                 Note: all symbols used in this expression must be defined in the price feed map.
+   *                 Ex: "(USDETH + COMPUSD) / COMPUSD"
+   * @param {number} decimals decimals to use in the price output.
+   */
+  constructor(priceFeedMap, expression, decimals = 18) {
+    super();
+    this.expressionCode = math.parse(expression).compile();
+    this.priceFeedMap = priceFeedMap;
+    this.decimals = decimals;
+  }
+
+  async getHistoricalPrice(time) {
+    const historicalPrices = {};
+    const errors = [];
+    await Promise.all(
+      Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+        const price = await pf.getHistoricalPrice(time).catch(err => errors.push(err));
```

Comment:
why accumulate errors here rather than throwing on the first one? you cant continue if you have a single error anyway

Reply:
This is the way we do this in the medianizer feed, so I just continued the pattern. The idea is that _all_ lower-level feed errors should be bubbled up for the purpose of logging in case multiple are failing. This has been useful for the complex stablespread feeds.